Donate media playback information to Screen Time

jesxilin opened the pull request:
#### a6aba3e394b3b74f9558859bf64de9d02c2923b6
<pre>
Donate media playback information to Screen Time
<a href="https://bugs.webkit.org/show_bug.cgi?id=287168">https://bugs.webkit.org/show_bug.cgi?id=287168</a>
<a href="https://rdar.apple.com/140439198">rdar://140439198</a>

Reviewed by Aditya Keerthi.

Add support for URLIsPictureInPicture and URLIsPlayingVideo on STWebpageController.
Add API tests for URLIsPlayingVideo and URLIsPictureInPicture.
Currently, there is only a mac test for URLIsPictureInPicture. This is because
TestWebKitAPI is not a UI application so can&apos;t actually go into PiP.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::setURLIsPictureInPictureForScreenTime):
(WebKit::PageClientImplCocoa::setURLIsPlayingVideoForScreenTime):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::hasVideoInPictureInPictureDidChange):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::setURLIsPictureInPictureForScreenTime):
(WebKit::PageClient::setURLIsPlayingVideoForScreenTime):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasVideoInPictureInPictureDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, URLIsPlayingVideo)):
(-[STPictureInPictureUIDelegate _webView:hasVideoInPictureInPictureDidChange:]):
(-[STPictureInPictureUIDelegate userContentController:didReceiveScriptMessage:]):
(TEST(ScreenTime, URLIsPictureInPictureMacos)):

Canonical link: <a href="https://commits.webkit.org/290235@main">https://commits.webkit.org/290235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17e9d6b87e1750dee783de4989178add2ac9d660

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8849 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94310 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40085 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17157 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68820 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26490 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92326 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7088 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81074 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49180 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6834 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35459 "Found 60 new test failures: compositing/animation/repaint-after-clearing-shared-backing.html editing/undo/redo-reapply-edit-command-crash.html fast/repaint/switch-overflow-vertical-rl.html fast/workers/dedicated-worker-lifecycle.html http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-script-in-iframe.html imported/w3c/web-platform-tests/content-security-policy/reporting/report-original-url-on-mixed-content-frame.https.sub.html imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-self-collapsing-nested.html imported/w3c/web-platform-tests/css/css-contain/contain-paint-containing-block-fixed-001.html imported/w3c/web-platform-tests/css/css-grid/grid-model/grid-inline-float-001.html imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/alignment/masonry-align-content-003.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39192 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96139 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16504 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77689 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76988 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21418 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20019 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9654 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14014 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16518 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21829 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16259 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19710 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13400,6 +13409,11 @@ void WebPageProxy::updatePlayingMediaDidChange(MediaProducerMediaStateFlags newS
+
+#if ENABLE(SCREEN_TIME)
+    if (oldState.contains(MediaProducerMediaState::IsPlayingVideo) != newState.contains(MediaProducerMediaState::IsPlayingVideo))
+    m_pageClient->setScreenTimeIsPlayingVideo(newState.contains(MediaProducerMediaState::IsPlayingVideo));
```
```suggestion
    if (oldState.contains(MediaProducerMediaState::IsPlayingVideo) != newState.contains(MediaProducerMediaState::IsPlayingVideo))
        m_pageClient->setScreenTimeIsPlayingVideo(newState.contains(MediaProducerMediaState::IsPlayingVideo));
```

lilyspiniolas commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
```diff
@@ -133,6 +133,8 @@ class PageClientImplCocoa : public PageClient {
     void installScreenTimeWebpageController() final;
     void didChangeScreenTimeWebpageControllerURL() final;
+    void setScreenTimePictureInPicture(bool) final;
+    void setScreenTimeIsPlayingVideo(bool) final;
```
nit: These method names sound like screen time itself is in Picture in Picture or playing video rather than a URL. I would update them to make it more clear.

pxlcoder replied:
Suggestion: `setURLIsPictureInPictureForScreenTime` / `setURLIsPlayingVideoForScreenTime`

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -401,6 +401,8 @@ - (void)_installScreenTimeWebpageController
         [_screenTimeWebpageController setSuppressUsageRecording:![_configuration websiteDataStore].isPersistent];
 
+        [_screenTimeWebpageController setURLIsPlayingVideo:NO];
+        [_screenTimeWebpageController setURLIsPictureInPicture:NO];
```
Unnecessary, as the default values are `NO`, and `_screenTimeWebpageController` is instantiated directly above.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+
+    [webView setUIDelegate:handler.get()];
+
+#if HAVE(TOUCH_BAR)
```
This is equivalent to `PLATFORM(MAC)` so you can remove the `#if`.

lilyspiniolas replied:
I might be wrong, but I don't think you need to check for HAVE(TOUCH_BAR) since the test is already enclosed with a platform check for Mac

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    [webView _forceRequestCandidates];
+#endif
+
+    RetainPtr<NSWindow> window = adoptNS([[NSWindow alloc] initWithContentRect:[webView frame] styleMask:NSWindowStyleMaskBorderless backing:NSBackingStoreBuffered defer:NO]);
```
```suggestion
    RetainPtr window = adoptNS([[NSWindow alloc] initWithContentRect:[webView frame] styleMask:NSWindowStyleMaskBorderless backing:NSBackingStoreBuffered defer:NO]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    [[window contentView] addSubview:webView.get()];
+    [window makeKeyAndOrderFront:nil];
+
+    NSURLRequest *request = [NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"PictureInPictureDelegate" withExtension:@"html"]];
```
```suggestion
    RetainPtr request = [NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"PictureInPictureDelegate" withExtension:@"html"]];
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    hasVideoInPictureInPictureValue = false;
+    hasVideoInPictureInPictureCalled = false;
+
+#if HAVE(TOUCH_BAR) && ENABLE(WEB_PLAYBACK_CONTROLS_MANAGER)
```
This is always true on `PLATFORM(MAC)`, so you can remove the `#if` and the code under the `#else`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    // Wait for PIPAgent to launch, or it won't call -pipDidClose: callback.
+    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
+
+#if HAVE(TOUCH_BAR) && ENABLE(WEB_PLAYBACK_CONTROLS_MANAGER)
```
Same comment.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    [webView _togglePictureInPicture];
+#else
+    [webView mouseDown:event];
+#endif
```
```suggestion
[webView _togglePictureInPicture];
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,111 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    EXPECT_FALSE([[webView _screenTimeWebpageController] URLIsPlayingVideo]);
+}
+
+@interface STPictureInPictureUIDelegate : NSObject <WKUIDelegate, WKScriptMessageHandler>
+@end
+
+@implementation STPictureInPictureUIDelegate
```
Move this under the  `#if PLATFORM(MAC)` if we end up unable to write an iOS test.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -349,6 +349,25 @@
+        return;
+
+    [screenTimeWebpageController setURLIsPlayingVideo:value];
+}
```
Do we need to skip setting this for non-persistent datastores? Or is that already handled by not donating the URL itself in that scenario.

jesxilin replied:
Just tested, it is already handled by not donating URL (`setSuppressUsageRecording`).

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -774,7 +774,7 @@ - (BOOL)prefersStatusBarHidden
     if (!page)
         return;
-    page->uiClient().hasVideoInPictureInPictureDidChange(page.get(), value);
+    m_page->hasVideoInPictureInPictureDidChange(value);
```
```suggestion
    page->hasVideoInPictureInPictureDidChange(value);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -32,15 +32,20 @@
 
 static void *blockedStateObserverChangeKVOContext = &blockedStateObserverChangeKVOContext;
 static bool stateDidChange = false;
+static bool recievedLoadMessage = false;
```
```suggestion
static bool receivedLoadMessage = false;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -32,15 +32,20 @@
 static bool stateDidChange = false;
+static bool recievedLoadMessage = false;
+static bool hasVideoInPictureInPictureValue;
+static bool hasVideoInPictureInPictureCalled;
```
```suggestion
static bool hasVideoInPictureInPictureValue = false;
static bool hasVideoInPictureInPictureCalled = false;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    EXPECT_FALSE([[webView _screenTimeWebpageController] URLIsPlayingVideo]);
+}
+
+#if PLATFORM(MAC)
```
```suggestion
#if PLATFORM(MAC)

```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+
+TEST(ScreenTime, URLIsPictureInPictureMacos)
+{
+    RetainPtr<WKWebViewConfiguration> configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+TEST(ScreenTime, URLIsPictureInPictureMacos)
+{
+    RetainPtr<WKWebViewConfiguration> configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    [configuration preferences]._allowsPictureInPictureMediaPlayback = YES;
```
```suggestion
    [configuration preferences]._allowsPictureInPictureMediaPlayback = YES;
    
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+{
+    RetainPtr<WKWebViewConfiguration> configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    [configuration preferences]._allowsPictureInPictureMediaPlayback = YES;
+    RetainPtr<STPictureInPictureUIDelegate> handler = adoptNS([[STPictureInPictureUIDelegate alloc] init]);
```
```suggestion
    RetainPtr handler = adoptNS([[STPictureInPictureUIDelegate alloc] init]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    [configuration preferences]._allowsPictureInPictureMediaPlayback = YES;
+    RetainPtr<STPictureInPictureUIDelegate> handler = adoptNS([[STPictureInPictureUIDelegate alloc] init]);
+    [[configuration userContentController] addScriptMessageHandler:handler.get() name:@"pictureInPictureChangeHandler"];
+    RetainPtr webView = webViewForScreenTimeTests(configuration.get());
```
```suggestion

    RetainPtr webView = webViewForScreenTimeTests(configuration.get());
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -331,4 +336,98 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
+    EXPECT_FALSE(hasVideoInPictureInPictureValue);
+    EXPECT_FALSE([[webView _screenTimeWebpageController] URLIsPictureInPicture]);
+}
+#endif
```
```suggestion
}

#endif

```

jesxilin committed:
Donate media playback information to Screen Time
https://bugs.webkit.org/show_bug.cgi?id=287168
rdar://140439198

Reviewed by Aditya Keerthi.

Add support for URLIsPictureInPicture and URLIsPlayingVideo on STWebpageController.
Add API tests for URLIsPlayingVideo and URLIsPictureInPicture.
Currently, there is only a mac test for URLIsPictureInPicture. This is because
TestWebKitAPI is not a UI application so can't actually go into PiP.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::setURLIsPictureInPictureForScreenTime):
(WebKit::PageClientImplCocoa::setURLIsPlayingVideoForScreenTime):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::hasVideoInPictureInPictureDidChange):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::setURLIsPictureInPictureForScreenTime):
(WebKit::PageClient::setURLIsPlayingVideoForScreenTime):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasVideoInPictureInPictureDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, URLIsPlayingVideo)):
(-[STPictureInPictureUIDelegate _webView:hasVideoInPictureInPictureDidChange:]):
(-[STPictureInPictureUIDelegate userContentController:didReceiveScriptMessage:]):
(TEST(ScreenTime, URLIsPictureInPictureMacos)):

Canonical link: https://commits.webkit.org/290235@main
